### DISCUSSION TITLE
Custom Eclipse Plugin compatibility issues due to changes in EclipseMavenLauncher class (IMavenLauncher Interface)

### DISCUSSION THREAD

**Discussion by kamal-kaur04** [eclipse-m2e/m2e-core#1365] [Q&A]
I'm currently creating a eclipse plugin wherein I'm launching a maven command by instantiating EclipseMavenLauncher class and calling runMaven method -

Now the method signature has changed as :

```
public CompletableFuture<?> runMaven(File basedir, String goals, Map<String, String> mavenProperties,
      boolean interactive)
```

Earlier, it used to accept userProperties as Properties class type. Now, if I try to install the plugin, execution fails on older version of eclipse. I tried to bundle the m2e plugin also. But it then ask to downgrade Eclipse.

Anything I'm doing wrong or how can I resolve this incompatibility? Any help is much appreciated. Thanks.

**Comment by laeubi** [comment]
`org.eclipse.m2e.core.internal.launch.IMavenLauncher` is an internal interface and not an extension plugin API, so you should either be prepared for API changes (that is your plugin is bound to a specific M2Eclipse version) or not use that interface.

If you have a strong use-case and require to be compatible, you might want to describe your case and ask for the interface to become a public API.

**Comment by kamal-kaur04** [comment]
Okay, I'll explain the use-case.

I'm using EclipseMavenLauncher class instance to execute **archetype generation command**, it uses Eclipse bundled Maven version and runs the command on the Eclipse console.

I could not execute command on the Terminal instead as it would need maven home to be present on the CLI config. Hence, I reused the EclipseMavenLauncher to execute maven command. I'm not sure if there is another way to do so, Is there any other right approach?

**Reply by laeubi** [reply]
Do you require a console for interactive input? If you you can use `IMaven` to execute maven like commands.

**Reply by kamal-kaur04** [reply]
> Do you require a console for interactive input? If you you can use `IMaven` to execute maven like commands.

No, interactive input is not required. It's just that the console needs to invoke for generation failures and respective logs.

**Reply by laeubi** [reply]
If you need the console output, I think there is no good way to archive this otherwise. The API should be quite stable now, so maybe the easiest solution would be to simply drop support for older eclipse releases at this point and simply adapt to the new method (users of older eclipse can still use previous releases of your plugin).
